### PR TITLE
[addons][pvr] add independent instance settings way

### PR DESCRIPTION
**Pull request by AlwinEsch**
## Description

The background to this request is to make PVR add-ons usable multiple times.

This is then done using an `instance-settings.xml` file and instance handling adapted in the addon.

## Screenshots (if appropriate):

Bildschirmfoto vom 2021-12-17 04-03-51
Bildschirmfoto vom 2022-01-08 15-31-33

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/settings/lib/SettingSection.h
Comment (Montellese) [inline xbmc/settings/lib/SettingSection.h:125]: `AddGroupToFront()`

**Review by Montellese — DISMISSED** [review]
I was wondering if it would make more sense to introduce a `CAddonInstance` class which holds the settings of a specific instance to avoid having to add the `instance` parameter to almost all methods in `IAddon` and `CAddon`. Then we would only have to add a `GetAddonInstance(instance)` method. Furthermore `CAddonInstance` could even also implement `IAddon` and could either return itself when `GetAddonInstance(instance)` is called with the proper `instance` or `nullptr` if the `instance` doesn't match.

It might also be a performance improvement in all those places where you have to call multiple methods of `IAddon` and do the instance lookup every time.

File: xbmc/pvr/addons/PVRClients.cpp
Comment (AlwinEsch) [reply xbmc/pvr/addons/PVRClients.cpp:237]: Yes, was come by change of his parent class, before PVR becomed it, as last commit was the support added to PVR and where it is needed, the `return StopClient(addonId, instance, true);` below..

**Comment by AlwinEsch** [comment]
> I was wondering if it would make more sense to introduce a `CAddonInstance` class which holds the settings of a specific instance to avoid having to add the `instance` parameter to almost all methods in `IAddon` and `CAddon`. Then we would only have to add a `GetAddonInstance(instance)` method. Furthermore `CAddonInstance` could even also implement `IAddon` and could either return itself when `GetAddonInstance(instance)` is called with the proper `instance` or `nullptr` if the `instance` doesn't match.
> 
> It might also be a performance improvement in all those places where you have to call multiple methods of `IAddon` and do the instance lookup every time.

Also considered putting it separately and created an `IAddonSettings` class which should then be given by CAddon and IAddonInstanceHandler.

Just thought then to leave it in existing code so related stuff (GUI) has to deal with just one class. In the GUI it is very limited to IAddon in `CGUIDialogAddonSettings::ShowForAddon` and the other functions.
But think about it again, maybe I'll think of something else.

**Comment by AlwinEsch** [comment]
Maybe easier possible to manage, if mostly only `CAddonInfo` becomes used and CAddon only on related processes where does his own works.

File: xbmc/addons/AddonStatusHandler.cpp
Comment (ksooo) [reply xbmc/addons/AddonStatusHandler.cpp:46]: > requested from @ksooo

 instance@addon readsnatural, imo - pretty much like user@host. Initially Alwin came up with addon@instance, which feels like wrong order to me.

File: xbmc/addons/Addon.cpp
Comment (ksooo) [reply xbmc/addons/Addon.cpp:88]: The name Alwin originally came up with was not reflecting what the implementation does. But the (quite unusual) name I suggested imo does.

File: xbmc/addons/AddonEvents.h
Comment (ksooo) [reply xbmc/addons/AddonEvents.h:11]: Self-contained header files, please. So yes, put it back.

File: xbmc/addons/AddonEvents.h
Comment (ksooo) [reply xbmc/addons/AddonEvents.h:48]: add-on is correct. In German we use Addon (without the dash).

File: xbmc/addons/binary-addons/AddonInstanceHandler.h
Comment (ksooo) [reply xbmc/addons/binary-addons/AddonInstanceHandler.h:27]: > Now we have instance, instanceID and instanceNumber. Please pick one and keep it consistent.

+1 for that

File: xbmc/addons/addoninfo/AddonInfo.cpp
Comment (ksooo) [reply xbmc/addons/addoninfo/AddonInfo.cpp:157]: Haha, I have no hard feelings about this.

File: xbmc/addons/addoninfo/AddonInfo.cpp
Comment (ksooo) [reply xbmc/addons/addoninfo/AddonInfo.cpp:281]: > Is this path also used somewhere else? Maybe it can be put into a constant?

@AlwinEsch exactly what I suggested - do you remember?

File: xbmc/addons/AddonStatusHandler.h
Comment (ksooo) [reply xbmc/addons/AddonStatusHandler.h:33]: > Was requested by @ksooo

 Really? Ooops.

Comment (ksooo) [reply xbmc/addons/AddonStatusHandler.h:33]: I'm just not a big fan of the "copy and move idiom". This can be discussed back and forth. Do it as you like, but either copy & move or const ref & no move.

File: addons/resource.language.en_gb/resources/strings.po
Comment (ksooo) [reply addons/resource.language.en_gb/resources/strings.po:4893]: > request of @ksooo

"instance" is far to technical imo. No end user will understand what this means. I consider "configuration" not ideal and I'm totally open for other suggestions.

File: xbmc/addons/addoninfo/AddonInfo.cpp
Comment (AlwinEsch) [reply xbmc/addons/addoninfo/AddonInfo.cpp:281]: I remember :smile:, thats why I mean to make on separate request. I'm normally not like to have two different parts within one request.

The best is to make this format only one time on one place.
Need to confirm that nothing goes wrong with repo content and not bring in silent problems by change of this.
There are also other parts where maybe can moved to CAddonInfo.

File: xbmc/addons/binary-addons/AddonInstanceHandler.cpp
Comment (Montellese) [reply xbmc/addons/binary-addons/AddonInstanceHandler.cpp:25]: IMO we shouldn't reduce the length and meaning of variable names and default values just to work around `clang-format`.

File: xbmc/settings/lib/SettingSection.h
Comment (Montellese) [reply xbmc/settings/lib/SettingSection.h:125]: Yeah the STL has it all: `begin()` and `end()`, `front()` and `back()` but the most matching name here would be `push_front()` like in `std::list` and other containers. That's why I would to with `AddGroupToFront()`. It also reads more naturally IMO.

**Comment by AlwinEsch** [comment]
Is rebased and updated/fixed few parts inside.

File: xbmc/addons/AddonStatusHandler.h
Comment (AlwinEsch) [reply xbmc/addons/AddonStatusHandler.h:33]: About the `const std::string& message`, maybe it makes sense to remove complete?
As reason it nowhere becomes set.

**Review by Montellese — COMMENTED** [review]
Unfortunately I don't have the time to look at the changes since my last review.

**Comment by phunkyfish** [comment]
Jenkins build this please

File: addons/resource.language.en_gb/resources/strings.po
Comment (garbear) [reply addons/resource.language.en_gb/resources/strings.po:4898]: The best synonym IMO: "setup". perfectly describes "configuration" and is way simpler. I even created a PR to change 'configuration' to "setup": https://github.com/xbmc/xbmc/pull/11654

**Comment by ksooo** [comment]
@AlwinEsch according to the discussion in this PR there are a couple of changes you agreed to make before merging this PR.

File: xbmc/addons/addoninfo/AddonInfo.cpp
Comment (AlwinEsch) [reply xbmc/addons/addoninfo/AddonInfo.cpp:154]: As them as static function not possible, static used to allow also outside.

File: xbmc/addons/addoninfo/AddonInfo.cpp
Comment (AlwinEsch) [reply xbmc/addons/addoninfo/AddonInfo.cpp:282]: In another request becomes changed.

File: xbmc/addons/AddonStatusHandler.h
Comment (AlwinEsch) [reply xbmc/addons/AddonStatusHandler.h:33]: In another request becomes the unused string removed.

File: xbmc/addons/addoninfo/AddonInfo.cpp
Comment (AlwinEsch) [reply xbmc/addons/addoninfo/AddonInfo.cpp:29]: With C++17 it can be becomed used as `std::string_view` and change it to them.

**Comment by AlwinEsch** [comment]
Is updated and should now includes most changes about the notes, thanks a lot about the improvement tips!

Has done now fast on rebase and still need to check the add-ons about.

**Comment by AlwinEsch** [comment]
The first commits relates to https://github.com/xbmc/xbmc/pull/21739 and will go away from here.

**Comment by AlwinEsch** [comment]
@ksooo @Montellese as the other was become in, is them here cleaned up and ready for review and tests.

@ksooo Maybe you can test with your pvr backends, have for now only tested with https://github.com/kodi-pvr/pvr.demo/pull/123

File: xbmc/addons/Addon.h
Comment (AlwinEsch) [inline xbmc/addons/Addon.h:430]: Think to add about here on another following request an change of bool to enum, like now a bit confusing and can be maybe by mistake become id instead of bool.

E.g.:
```cpp
enum class SettingsForceLoad
{
  YES,
  NO
};

enum class SettingsLoadUserPart
{
  YES,
  NO
};

bool LoadSettings(SettingsForceLoad bForce, SettingsLoadUserPart loadUserSettings, AddonInstanceId id = ADDON_SETTINGS_ID);
```

@ksooo What you think about?

Comment (enen92) [reply xbmc/addons/Addon.h:430]: Better to avoid YES and NO as enum values as they seem to already exist in Darwin and have been a cause for past issues. @howie-f should know better (I think CHOICE_YES and CHOICE_NO has been used instead)

Comment (AlwinEsch) [reply xbmc/addons/Addon.h:430]: Also relates to an "enum class"?

Comment (enen92) [reply xbmc/addons/Addon.h:430]: Yes, see https://github.com/xbmc/xbmc/pull/20925

Comment (ksooo) [reply xbmc/addons/Addon.h:430]: @AlwinEsch generally I like the idea, but may I ask you not to add more stuff to this PR, please? It is already large and difficult to review. So, I would appreciate if you put this kind of cosmetic enhancements not needed for the addon-API changes in a follow-up PR. For now, let's focus on getting this PR in shape, without adding more to it.

**Comment by ksooo** [comment]
> @ksooo Maybe you can test with your pvr backends, have for now only tested with https://github.com/kodi-pvr/pvr.demo/pull/123

Regression tests ensuring ols functionality is not broken can be done by anybody. To test the new functionality, I would need to do larger changes to at least one pvr add-on and I'm afraid I will not find the time for that anytime soon.

File: xbmc/pvr/addons/PVRClients.cpp
Comment (ksooo) [inline xbmc/pvr/addons/PVRClients.cpp:40]: Shouldn't the string to hash read "instance@addonID"? Current implementation produces for non-empty instance "@instanceaddonID"...

File: xbmc/pvr/channels/PVRChannelsPath.cpp
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannelsPath.cpp:96]: You implemented the new path so that instanceid is always present. So, the comment should read: "// pvr://channels/(tv|radio)/<groupname>/<instanceid>@<addonid>_<channeluid>.pvr"

File: xbmc/addons/Addon.h
Comment (AlwinEsch) [reply xbmc/addons/Addon.h:430]: Yes, not thought about for here (see the "here on another following request" above), only as the place visible here I thought to ask about 😀.

**Review by ksooo — COMMENTED** [review]
minors only

**Comment by AlwinEsch** [comment]
Can try to change pvr.zattoo about, as this already allow different web hosts in his settings.

File: xbmc/addons/Addon.h
Comment (ksooo) [reply xbmc/addons/Addon.h:430]: There is much more room for improving several function signatures, like getting rid of bool parameters, replacing them with speaking enum values etc. But all this is really for another PR.

**Comment by ksooo** [comment]
I adapted pvr.hts (PR to follow). After ironing out some glitches in this PR (now in), the new feature works like a charm. 

I will merge the PR once Jenkins is green.

BTW: This is all compatible Add- on API changes, so now new add-on release needed.

**Review by ksooo — APPROVED** [review]
A very cool feature, well implemented. :-)

**Comment by phunkyfish** [comment]
Excellent, great to see this going in. Now just to find the time to do the same for pvr.vuplus and pvr.iptvsimple! Thanks @AlwinEsch !!!

File: addons/resource.language.en_gb/resources/strings.po
Comment (gade01) [inline addons/resource.language.en_gb/resources/strings.po:4915]: @AlwinEsch I k ow this PR already went in, but there's a missing `mscstr ""` at line 4916.

Comment (ksooo) [reply addons/resource.language.en_gb/resources/strings.po:4915]: I will open a PR with a fix.

**Pull request merged**
